Bugfix for saving empty multi selects or radio fields.

discoinfiltrator opened the pull request:
A few fixes here:

- $data should be an array, as that's how it's used in save_walk_children
- save_walk_children only has two function parameters, so removing $this->fm->data_id from the call
- Since we're relying on the POST array, we need to check our fields for multi-selects and radio buttons (which when empty are not passed in the POST array)

Relevant GitHub issue is:

https://github.com/alleyinteractive/wordpress-fieldmanager/issues/685

dlh01 commented on php/context/class-fieldmanager-context-storable.php:
```diff
@@ -52,9 +52,18 @@ protected function save( $data = null ) {
+				$data = isset( $_POST[ $this->fm->name ] ) ? $_POST[ $this->fm->name ] : array(); // WPCS: input var okay. CSRF okay. Sanitization okay.
+
+				// Multiple select and radio fields with no values chosen are left out of the post request altogether.
+				foreach ( $this->fm->children as $child ) {
```
Because `serialize_data` is technically a property on `Fieldmanager_Field`, is it safe to assume that `$this->fm` is going to have a `children` property?

dlh01 commented on php/context/class-fieldmanager-context-storable.php:
```diff
@@ -52,9 +52,18 @@ protected function save( $data = null ) {
+				// Multiple select and radio fields with no values chosen are left out of the post request altogether.
+				foreach ( $this->fm->children as $child ) {
+					$is_multi_select = ( 'select' == $child->field_class ) && ( true === $child->multiple );
+					$is_radio = ( 'radio' == $child->field_class );
```
Perhaps this could use `instanceof` instead of the `field_class` property, which is public. See 878f9e55a19882444eb0a4fa0f8108e192a6a89f.

dlh01 commented on php/context/class-fieldmanager-context-storable.php:
```diff
@@ -52,9 +52,18 @@ protected function save( $data = null ) {
+				foreach ( $this->fm->children as $child ) {
+					$is_multi_select = ( 'select' == $child->field_class ) && ( true === $child->multiple );
+					$is_radio = ( 'radio' == $child->field_class );
+					if ( ( $is_multi_select || $is_radio ) && ! isset( $data[ $child->name ] ) && ! $child->skip_save ) {
```
It looks like `skip_save` is also checked in `\Fieldmanager_Context_Storable::save_field()`. Is it needed here?

discoinfiltrator commented:
Hey David - much appreciated for taking a look. I'll add tests and address your comments over the next week or so.

discoinfiltrator commented:
Hey @dlh01 - I've updated my code based on your feedback, and added a test case (forgive all the dumb commits).  The second assertSame() fails without my fix.

I tested the hidden input solution for select inputs and it does work. I tried modifying class-fieldmanager-select.php like:

```
return sprintf(
	'<input type="hidden" name="%2$s" value="" /><select class="%1$s" name="%2$s" id="%3$s" %4$s>%5$s</select>',
	esc_attr( implode( ' ', $select_classes ) ),
	esc_attr( $this->get_form_name( $do_multiple ) ),
	esc_attr( $this->get_element_id() ),
	$this->get_element_attributes(),
	$opts
);
```

I don't have a preference here. If you decide to go with the hidden input solution, I don't think I can add a test case for it since it would require an end-to-end test (it doesn't look like these exist for this repo).

discoinfiltrator commented:
This PR seems to have been forgotten. CI builds are failing on files that were not modified as part of this work. This seems to be happening on other PRs as well. The main file I changed is passing:

```
No syntax errors detected in ./php/context/class-fieldmanager-context-storable.php
```

I have been using my change in production for some time now. Is there any way we can get this merged so I can update fieldmanager in my WP application without having to re-apply my changes?

dlh01 commented:
I am not, but I can try to give this another look soon.

discoinfiltrator commented:
@dlh01 , who would be a good person to review this? It has been some time and it would be nice to get merged.

dlh01 commented:
Perhaps @mboynes or @mslinnea could review?

discoinfiltrator commented:
@mboynes , @mslinnea , @dlh01 - is anybody able to help take a look at this? Or is there somebody else who can help? We've been using my fix in production for two years and it would be really nice to have it merged into the main repo.

discoinfiltrator commented:
@dlh01 @kevinfodness @mboynes @srtfisher what are the next steps for getting PRs merged? This one has been setting around for a long time and I'd love to help get it merged.

dlh01 commented:
Hey @discoinfiltrator — we're planning to work through some Fieldmanager updates in the next few weeks, including this PR. Hope to have more for you soon!

discoinfiltrator commented:
@dlh01 sounds great! For what it's worth, we've been using the code in this PR in production on a high traffic WP SaaS site with no issues.

dlh01 commented:
One year later 😬 we have an updated version of this Pull Request in #811, @discoinfiltrator. It includes some refinements  that narrow the scope a bit and performs a bit of other cleanup, but it's otherwise largely the same. Your feedback would be welcome!

dlh01 commented:
Superseded by #811, which was included in Fieldmanager 1.4.
